### PR TITLE
feat(o2k): add OIDC Security conversions

### PR DESCRIPTION
**Pull request by Tieske**
wip

### CODE REVIEW

File: docs/security-o2k.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+Within OpenAPI the security directives can be specified on the document root (the `OpenAPI` object). It can also be specified on the `Operation` object, in which case it will override the document level one.
+Specifically; it cannot be specified on the `path` object level, the level in between document and operation.
+
+## Format
+
+The `security` property is an array of `Security Requirement` objects.
+
+## Behaviour
+
+- The requirements listed are a logical `OR`; passing any one requirement listed is enough to be allowed access.
```
Comment (rspurgeon) [inline docs/security-o2k.md:21]: This does not describe the full ability as it relates to OAS and your examples below.   Both `OR` and `AND` configurations are possible, as you describe lower in this doc.

File: docs/security-o2k.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+- The default value is `[]`; an empty array
+- An empty array allows access without authentication
+- An empty array (on the `Operation` level) can override a non-empty array (document level)
+
+## Security Requirement
+
+A `security requirement` object lists (as properties) `security scheme` objects that MUST all be satisfied to allow access.
```
Comment (rspurgeon) [inline docs/security-o2k.md:28]: The OAS document does a good job of describing the AND / OR relationship here.  https://swagger.io/docs/specification/authentication/. I'm worried this documentation confuses matters as it's not as clear as what is written there.

File: docs/security-o2k.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+To authenticate a request must either pass:
+- `myOpenId` (with 'scope3') **AND** `myBasicAuth`, or
+- `myBasicAuth` **AND** `myKeyAuth`
+
+**NOTE**: the values (the arrays), contain scope names to use for `oauth2` and `openIdConnect` security schemes.
+(for all other types the array must be empty).
+
+## Security Scheme
+
+The `security scheme` object specifies a single security/authication mechanism to validate. The schemes are defined as properties on the `components.securitySchemes` object.
```
Comment (rspurgeon) [inline docs/security-o2k.md:47]:
```suggestion
The `security scheme` object specifies a single security/authentication mechanism to validate. The schemes are defined as properties on the `components.securitySchemes` object.
```

Generally, i'm not sure we need to rebuild the OAS documentation in our repo docs.  The document [here](https://swagger.io/docs/specification/authentication/) provides good explanation of OAS security specification design.

File: docs/security-o2k.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+Type | supported | Kong plugin
+-|-|-
+`http`| no |
+`apiKey`| no |
+`openIdConnect`| yes | `openid-connect` |
+`oauth2`| no |
+
+## Boolean logic
+
+No boolean AND/OR logic is supported. So a `security` directive can only have 1 `security requirement`, and with in that only a single `securityScheme`.
```
Comment (rspurgeon) [inline docs/security-o2k.md:80]:
```suggestion
No boolean AND/OR logic is supported. So a `security` directive can only have 1 `security requirement`, and within that only a single `securityScheme`.
```

File: docs/security-o2k.md (hunk 2 above)
Comment (Tieske) [reply docs/security-o2k.md:28]: Agreed.

thx for that link. Never seen before. The OAS spec itself is very terse on this subject. Hence this doc is for now mostly intended as a "thinking" doc. All OAS specifics we should probably remove (or add that link) before merging.

File: docs/security-o2k.md (hunk 1 above)
Comment (Tieske) [reply docs/security-o2k.md:21]: clarified, this only describes the `security` directive (which handles `OR`). The `securityRequirement` object defines the `AND`.

**Pull request merged**